eth/beacon/db/chain.py

```diff
@@ -289,61 +290,79 @@ def persist_block_chain(
+        )
+
     @classmethod
     def _persist_block_chain(
             cls,
             db: BaseDB,
             blocks: Iterable[BaseBeaconBlock],
             block_class: Type[BaseBeaconBlock]
     ) -> Tuple[Tuple[BaseBeaconBlock, ...], Tuple[BaseBeaconBlock, ...]]:
+        blocks_iterator = iter(blocks)
+
         try:
-            first_block = first(blocks)
+            first_block = first(blocks_iterator)
         except StopIteration:
             return tuple(), tuple()
-        else:
-            for parent, child in sliding_window(2, blocks):
-                if parent.root != child.parent_root:
-                    raise ValidationError(
-                        "Non-contiguous chain. Expected {} to have {} as parent but was {}".format(
-                            encode_hex(child.root),
-                            encode_hex(parent.root),
-                            encode_hex(child.parent_root),
-                        )
-                    )
 
-            is_genesis = first_block.parent_root == GENESIS_PARENT_HASH
-            if not is_genesis and not cls._block_exists(db, first_block.parent_root):
-                raise ParentNotFound(
-                    "Cannot persist block ({}) with unknown parent ({})".format(
-                        encode_hex(first_block.root), encode_hex(first_block.parent_root)))
+        is_genesis = first_block.parent_root == GENESIS_PARENT_HASH
+        if not is_genesis and not cls._block_exists(db, first_block.parent_root):
+            raise ParentNotFound(
+                "Cannot persist block ({}) with unknown parent ({})".format(
+                    encode_hex(first_block.root), encode_hex(first_block.parent_root)))
 
-            if is_genesis:
-                score = 0
-            else:
-                score = cls._get_score(db, first_block.parent_root)
+        if is_genesis:
+            score = 0
+        else:
+            score = cls._get_score(db, first_block.parent_root)
 
-        for block in blocks:
-            db.set(
-                block.root,
-                rlp.encode(block),
-            )
+        curr_block_head = first_block
+        db.set(
+            curr_block_head.root,
+            rlp.encode(curr_block_head),
+        )
+        cls._set_block_scores_to_db(db, curr_block_head)
 
-            # TODO: It's a stub before we implement fork choice rule
-            score = block.slot
+        orig_blocks_seq = concat([(first_block,), blocks_iterator])
+
+        for parent, child in sliding_window(2, orig_blocks_seq):
+            if parent.root != child.parent_root:
+                raise ValidationError(
+                    "Non-contiguous chain. Expected {} to have {} as parent but was {}".format(
+                        encode_hex(child.root),
+                        encode_hex(parent.root),
+                        encode_hex(child.parent_root),
+                    )
+                )
 
+            curr_block_head = child
```

Comment:
If I'm reading this correctly we *should* be able to drop this assignment and just use `child` in all of the places where `curr_block_head` is used from here through the end of the function, with the possible need to do some minor additional handling for the calls to `cls._set_as_canonical_chain_head`.

Reply:
That's right, though I'm not sure there is a better way to achieve this than what I already tried to do here. The biggest constraints we have here that prevented me from coming up with a better approach were -
1. The chain can be an iterator.
2. We want to loop only once

I'm open to ideas.